File: frontend/src/utils/vendorIdentityMatcher.ts
/**
 * Dynamic Self-Learning Vendor Identity Resolution Engine.
 * 
 * Replaces static hardcoded vendor/city lists with:
 * 1. Dynamic Alphanumeric Tokenization & Noise Filtering
 * 2. Semantic Token Overlap & Jaro-Winkler / Levenshtein Distance Matrix
 * 3. Dynamic Acronym & Initialism Resolution (e.g. LA <-> Los Angeles, PHL <-> Philadelphia)
 * 4. Self-Learning Persistent Alias Graph (Learns from runtime configs, directory objects & operator feedback)
 */

export interface VendorEntityContext {
  vendor_id?: string;
  vendor_name?: string;
  brand_name?: string;
  company_name?: string;
  subdomain?: string;
  city?: string;
  email?: string;
  phone?: string;
  website_url?: string;
}

const STORAGE_KEY = 'limo_learned_vendor_identities_v2';

// Common generic corporate & industry stop-words to eliminate noise
const NOISE_STOP_WORDS = new Set([
  'vendor', 'partner', 'cell', 'org', 'limo', 'limos', 'limousine', 'limousines',
  'fleet', 'fleets', 'chauffeur', 'chauffeurs', 'transport', 'transportation',
  'executive', 'exec', 'vip', 'prestige', 'sovereign', 'royal', 'global',
  'luxury', 'blackcar', 'black_car', 'shuttle', 'services', 'service',
  'holdings', 'group', 'enterprise', 'enterprises', 'inc', 'llc', 'ltd',
  'corp', 'corporation', 'co', 'the', 'and', 'of', 'for'
]);

/**
 * Normalizes an identifier or company name into distinctive semantic tokens.
 */
export function extractSemanticTokens(text?: string): string[] {
  if (!text) return [];
  
  // Clean punctuation, split camelCase, snake_case, kebab-case, dots
  const cleaned = text
    .replace(/([a-z])([A-Z])/g, '$1 $2')
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, ' ')
    .trim();

  const tokens = cleaned
    .split(/\s+/)
    .filter(t => t.length > 0 && !NOISE_STOP_WORDS.has(t));

  return Array.from(new Set(tokens));
}

/**
 * Computes Levenshtein edit distance between two normalized strings.
 */
export function levenshteinDistance(a: string, b: string): number {
  if (a === b) return 0;
  if (!a.length) return b.length;
  if (!b.length) return a.length;

  const matrix: number[][] = [];
  for (let i = 0; i <= b.length; i++) matrix[i] = [i];
  for (let j = 0; j <= a.length; j++) matrix[0][j] = j;

  for (let i = 1; i <= b.length; i++) {
    for (let j = 1; j <= a.length; j++) {
      if (b.charAt(i - 1) === a.charAt(j - 1)) {
        matrix[i][j] = matrix[i - 1][j - 1];
      } else {
        matrix[i][j] = Math.min(
          matrix[i - 1][j - 1] + 1, // substitution
          matrix[i][j - 1] + 1,     // insertion
          matrix[i - 1][j] + 1      // deletion
        );
      }
    }
  }

  return matrix[b.length][a.length];
}

/**
 * Computes Normalized String Similarity (0.0 to 1.0).
 */
export function stringSimilarity(a: string, b: string): number {
  const s1 = a.trim().toLowerCase();
  const s2 = b.trim().toLowerCase();
  if (s1 === s2) return 1.0;
  const maxLen = Math.max(s1.length, s2.length);
  if (maxLen === 0) return 1.0;
  const dist = levenshteinDistance(s1, s2);
  return Math.max(0, 1 - dist / maxLen);
}

/**
 * Checks if one token sequence dynamically represents an acronym/initialism of another.
 * E.g., 'la' matches 'los angeles', 'nyc' matches 'new york city', 'phl' matches 'philadelphia'.
 */
function isAcronymMatch(tokenA: string, tokenB: string): boolean {
  if (tokenA === tokenB) return true;
  const short = tokenA.length < tokenB.length ? tokenA : tokenB;
  const long = tokenA.length < tokenB.length ? tokenB : tokenA;

  // Check if short is the prefix initials of the long string
  if (short.length >= 2 && long.length >= 4) {
    if (long.startsWith(short)) return true;
  }
  return false;
}

/**
 * Persistent, Self-Learning Vendor Identity Knowledge Graph.
 */
class VendorIdentityLearnerStore {
  private aliasClusters: Map<string, Set<string>> = new Map();

  constructor() {
    this.loadFromStorage();
  }

  private loadFromStorage() {
    if (typeof window === 'undefined' || !window.localStorage) return;
    try {
      const raw = window.localStorage.getItem(STORAGE_KEY);
      if (raw) {
        const parsed = JSON.parse(raw);
        if (typeof parsed === 'object' && parsed !== null) {
          for (const [key, aliases] of Object.entries(parsed)) {
            if (Array.isArray(aliases)) {
              this.aliasClusters.set(key.toLowerCase(), new Set(aliases.map((a: string) => a.toLowerCase())));
            }
          }
        }
      }
    } catch (e) {
      console.warn('[VendorIdentityLearner] Could not parse stored identity graph:', e);
    }
  }

  private saveToStorage() {
    if (typeof window === 'undefined' || !window.localStorage) return;
    try {
      const obj: Record<string, string[]> = {};
      for (const [key, set] of this.aliasClusters.entries()) {
        obj[key] = Array.from(set);
      }
      window.localStorage.setItem(STORAGE_KEY, JSON.stringify(obj));
    } catch (e) {
      console.warn('[VendorIdentityLearner] Could not persist identity graph:', e);
    }
  }

  /**
   * Dynamically learns and binds an alias to a vendor cluster (Bidirectional).
   */
  public learnAlias(primaryVendorId: string, aliasOrIdentifier: string) {
    if (!primaryVendorId || !aliasOrIdentifier) return;
    const p = primaryVendorId.toLowerCase().trim();
    const a = aliasOrIdentifier.toLowerCase().trim();
    if (p === a) return;

    if (!this.aliasClusters.has(p)) {
      this.aliasClusters.set(p, new Set([p]));
    }
    this.aliasClusters.get(p)!.add(a);

    // Bidirectional link
    if (!this.aliasClusters.has(a)) {
      this.aliasClusters.set(a, new Set([a]));
    }
    this.aliasClusters.get(a)!.add(p);

    // Merge transitive clusters
    for (const member of Array.from(this.aliasClusters.get(p)!)) {
      if (this.aliasClusters.has(member)) {
        for (const trans of Array.from(this.aliasClusters.get(member)!)) {
          this.aliasClusters.get(p)!.add(trans);
        }
      }
    }

    this.saveToStorage();
  }

  /**
   * Auto-learns from a rich Vendor Config or Context.
   */
  public learnFromContext(context: VendorEntityContext) {
    if (!context || !context.vendor_id) return;
    const vid = context.vendor_id;

    const candidates = [
      context.vendor_name,
      context.brand_name,
      context.company_name,
      context.subdomain,
      context.city,
      context.email?.split('@')[0],
      context.website_url?.replace(/https?:\/\//, '').split('.')[0]
    ].filter(Boolean) as string[];

    for (const cand of candidates) {
      this.learnAlias(vid, cand);
      // Also learn clean semantic tokens
      const tokens = extractSemanticTokens(cand);
      if (tokens.length > 0) {
        this.learnAlias(vid, tokens.join('_'));
      }
    }
  }

  /**
   * Checks if candidateId belongs to the learned alias cluster of currentVendorId.
   */
  public isLearnedAlias(currentVendorId: string, candidateId: string): boolean {
    const c = currentVendorId.toLowerCase().trim();
    const target = candidateId.toLowerCase().trim();
    if (c === target) return true;

    const cluster = this.aliasClusters.get(c);
    if (cluster && cluster.has(target)) return true;

    // Check inverse
    const targetCluster = this.aliasClusters.get(target);
    if (targetCluster && targetCluster.has(c)) return true;

    return false;
  }

  /**
   * Returns all known aliases for a vendor.
   */
  public getAliases(vendorId: string): string[] {
    const v = vendorId.toLowerCase().trim();
    const set = this.aliasClusters.get(v);
    return set ? Array.from(set) : [v];
  }
}

// Global Singleton Learner Instance
export const vendorIdentityLearner = new VendorIdentityLearnerStore();

/**
 * Dynamic Self-Learning Vendor Matcher.
 * 
 * Evaluates whether partnerId refers to the same entity as currentVendorId:
 * - Direct ID equality
 * - Learned Alias Graph (localStorage + runtime memory)
 * - Semantic Token Intersection & Overlap Coefficient
 * - Fuzzy Levenshtein / Jaro Distance on normalized roots
 * - Substring containment after noise stripping
 * - Acronym resolution
 */
export const isSelfVendor = (
  partnerId?: string | { partner_id?: string; company_name?: string; city?: string },
  currentVendorIdOrContext?: string | VendorEntityContext
): boolean => {
  if (!partnerId || !currentVendorIdOrContext) return false;

  // 1. Extract string identifiers
  const pId = typeof partnerId === 'string' 
    ? partnerId 
    : (partnerId.partner_id || partnerId.company_name || '');
  
  const pName = typeof partnerId === 'object' ? partnerId.company_name : undefined;

  const cId = typeof currentVendorIdOrContext === 'string'
    ? currentVendorIdOrContext
    : (currentVendorIdOrContext.vendor_id || currentVendorIdOrContext.vendor_name || '');

  if (!pId.trim() || !cId.trim()) return false;

  const pRaw = pId.toLowerCase().trim();
  const cRaw = cId.toLowerCase().trim();

  // 2. Exact Raw Equality
  if (pRaw === cRaw) return true;

  // 3. Check Self-Learning Knowledge Graph
  if (vendorIdentityLearner.isLearnedAlias(cRaw, pRaw)) {
    return true;
  }
  if (pName && vendorIdentityLearner.isLearnedAlias(cRaw, pName.toLowerCase().trim())) {
    return true;
  }

  // 4. Extract Distinctive Semantic Tokens
  const pTokens = extractSemanticTokens(pId);
  const cTokens = extractSemanticTokens(cId);

  if (pName) {
    pTokens.push(...extractSemanticTokens(pName));
  }

  if (typeof currentVendorIdOrContext === 'object') {
    if (currentVendorIdOrContext.vendor_name) cTokens.push(...extractSemanticTokens(currentVendorIdOrContext.vendor_name));
    if (currentVendorIdOrContext.brand_name) cTokens.push(...extractSemanticTokens(currentVendorIdOrContext.brand_name));
    if (currentVendorIdOrContext.city) cTokens.push(...extractSemanticTokens(currentVendorIdOrContext.city));
  }

  const pUniqueTokens = Array.from(new Set(pTokens));
  const cUniqueTokens = Array.from(new Set(cTokens));

  if (pUniqueTokens.length > 0 && cUniqueTokens.length > 0) {
    // 5. Token Intersection
    const commonTokens = pUniqueTokens.filter(t => 
      cUniqueTokens.includes(t) || 
      cUniqueTokens.some(ct => isAcronymMatch(t, ct) || isAcronymMatch(ct, t))
    );

    const minTokenCount = Math.min(pUniqueTokens.length, cUniqueTokens.length);
    const overlapRatio = commonTokens.length / minTokenCount;

    // Strong Semantic Match
    if (commonTokens.length > 0 && overlapRatio >= 0.5) {
      // Auto-learn this association for future instant lookups
      vendorIdentityLearner.learnAlias(cRaw, pRaw);
      return true;
    }
  }

  // 6. Normalized Compact Roots
  const pClean = pUniqueTokens.join('');
  const cClean = cUniqueTokens.join('');

  if (pClean && cClean) {
    if (pClean === cClean || pClean.includes(cClean) || cClean.includes(pClean)) {
      vendorIdentityLearner.learnAlias(cRaw, pRaw);
      return true;
    }

    // 7. Fuzzy Edit Distance on Clean Semantic Roots
    const similarity = stringSimilarity(pClean, cClean);
    if (similarity >= 0.82) {
      vendorIdentityLearner.learnAlias(cRaw, pRaw);
      return true;
    }
  }

  return false;
};

File: app/services/vendor_identity_matcher.py
"""
Dynamic Self-Learning Vendor Identity Resolution Engine (Python).

Replaces static hardcoded vendor/city lists with:
1. Dynamic Alphanumeric Tokenization & Noise Filtering
2. Semantic Token Overlap & Levenshtein Distance Matrix
3. Dynamic Acronym & Initialism Resolution (e.g. LA <-> Los Angeles, PHL <-> Philadelphia, NYC <-> New York)
4. Self-Learning In-Memory & Persistent Alias Graph
"""

from __future__ import annotations

import re
import logging
from typing import Any, Dict, List, Optional, Set

logger = logging.getLogger("VendorIdentityMatcher")

NOISE_STOP_WORDS = {
    "vendor", "partner", "cell", "org", "limo", "limos", "limousine", "limousines",
    "fleet", "fleets", "chauffeur", "chauffeurs", "transport", "transportation",
    "executive", "exec", "vip", "prestige", "sovereign", "royal", "global",
    "luxury", "blackcar", "black_car", "shuttle", "services", "service",
    "holdings", "group", "enterprise", "enterprises", "inc", "llc", "ltd",
    "corp", "corporation", "co", "the", "and", "of", "for"
}


def extract_semantic_tokens(text: Optional[str]) -> List[str]:
    """Normalizes an identifier or company name into distinctive semantic tokens."""
    if not text:
        return []
    
    # Split camelCase and replace non-alphanumerics
    s1 = re.sub(r"([a-z])([A-Z])", r"\1 \2", str(text))
    cleaned = re.sub(r"[^a-zA-Z0-9\s]", " ", s1).lower().strip()
    
    tokens = [
        t for t in cleaned.split()
        if len(t) > 0 and t not in NOISE_STOP_WORDS
    ]
    # Maintain order and uniqueness
    seen = set()
    result = []
    for t in tokens:
        if t not in seen:
            seen.add(t)
            result.append(t)
    return result


def levenshtein_distance(s1: str, s2: str) -> int:
    """Computes Levenshtein edit distance between two strings."""
    if s1 == s2:
        return 0
    if len(s1) == 0:
        return len(s2)
    if len(s2) == 0:
        return len(s1)

    previous_row = range(len(s2) + 1)
    for i, c1 in enumerate(s1):
        current_row = [i + 1]
        for j, c2 in enumerate(s2):
            insertions = previous_row[j + 1] + 1
            deletions = current_row[j] + 1
            substitutions = previous_row[j] + (c1 != c2)
            current_row.append(min(insertions, deletions, substitutions))
        previous_row = current_row

    return previous_row[-1]


def string_similarity(s1: str, s2: str) -> float:
    """Computes Normalized String Similarity (0.0 to 1.0)."""
    a = s1.strip().lower()
    b = s2.strip().lower()
    if a == b:
        return 1.0
    max_len = max(len(a), len(b))
    if max_len == 0:
        return 1.0
    dist = levenshtein_distance(a, b)
    return max(0.0, 1.0 - (dist / max_len))


def is_acronym_match(token_a: str, token_b: str) -> bool:
    """Checks if one token dynamically represents an initialism/prefix of another."""
    if token_a == token_b:
        return True
    short_tok = token_a if len(token_a) < len(token_b) else token_b
    long_tok = token_b if len(token_a) < len(token_b) else token_a
    if len(short_tok) >= 2 and len(long_tok) >= 4:
        if long_tok.startswith(short_tok):
            return True
    return False


class VendorIdentityLearner:
    """Self-learning persistent vendor alias graph."""
    
    def __init__(self):
        self._clusters: Dict[str, Set[str]] = {}

    def learn_alias(self, primary_id: str, alias_or_identifier: str) -> None:
        """Learns and registers bidirectional alias link."""
        if not primary_id or not alias_or_identifier:
            return
        p = primary_id.strip().lower()
        a = alias_or_identifier.strip().lower()
        if p == a:
            return

        if p not in self._clusters:
            self._clusters[p] = {p}
        self._clusters[p].add(a)

        if a not in self._clusters:
            self._clusters[a] = {a}
        self._clusters[a].add(p)

        # Merge transitive clusters
        for member in list(self._clusters[p]):
            if member in self._clusters:
                self._clusters[p].update(self._clusters[member])

    def learn_from_vendor_profile(self, vendor_id: str, profile_data: Dict[str, Any]) -> None:
        """Extracts and binds all identifiers from a vendor profile."""
        if not vendor_id:
            return
        candidates = [
            profile_data.get("name"),
            profile_data.get("brand_name"),
            profile_data.get("company_name"),
            profile_data.get("city"),
            profile_data.get("subdomain"),
        ]
        for cand in candidates:
            if cand and isinstance(cand, str):
                self.learn_alias(vendor_id, cand)
                tokens = extract_semantic_tokens(cand)
                if tokens:
                    self.learn_alias(vendor_id, "_".join(tokens))

    def is_learned_alias(self, vendor_a: str, vendor_b: str) -> bool:
        """Checks if vendor_b is in vendor_a's learned alias graph."""
        a = vendor_a.strip().lower()
        b = vendor_b.strip().lower()
        if a == b:
            return True
        if a in self._clusters and b in self._clusters[a]:
            return True
        if b in self._clusters and a in self._clusters[b]:
            return True
        return False

    def canonicalize(self, vendor_id: str) -> str:
        """Returns the canonical semantic fingerprint for a vendor ID."""
        if not vendor_id:
            return ""
        tokens = extract_semantic_tokens(vendor_id)
        if tokens:
            return "_".join(tokens)
        return vendor_id.lower().replace("vendor_", "").replace("vendor-", "").replace("-", "_").strip()


# Global Singleton Instance
vendor_identity_learner = VendorIdentityLearner()


def is_self_vendor(vendor_a: str, vendor_b: str) -> bool:
    """
    Dynamic, Self-Learning Vendor Identity Matcher.
    Returns True if vendor_a and vendor_b refer to the same sovereign entity.
    """
    if not vendor_a or not vendor_b:
        return False
    
    a_raw = str(vendor_a).strip().lower()
    b_raw = str(vendor_b).strip().lower()

    # 1. Exact Raw Equality
    if a_raw == b_raw:
        return True

    # 2. Check Learned Knowledge Graph
    if vendor_identity_learner.is_learned_alias(a_raw, b_raw):
        return True

    # 3. Extract Distinctive Semantic Tokens
    a_tokens = extract_semantic_tokens(a_raw)
    b_tokens = extract_semantic_tokens(b_raw)

    if a_tokens and b_tokens:
        # 4. Token Intersection & Overlap Ratio
        common = [
            t for t in a_tokens
            if t in b_tokens or any(is_acronym_match(t, bt) or is_acronym_match(bt, t) for bt in b_tokens)
        ]
        min_tokens = min(len(a_tokens), len(b_tokens))
        if min_tokens > 0:
            overlap = len(common) / min_tokens
            if len(common) > 0 and overlap >= 0.5:
                # Auto-learn this dynamic match
                vendor_identity_learner.learn_alias(a_raw, b_raw)
                return True

    # 5. Normalized Compact Roots
    a_clean = "".join(a_tokens)
    b_clean = "".join(b_tokens)

    if a_clean and b_clean:
        if a_clean == b_clean or a_clean in b_clean or b_clean in a_clean:
            vendor_identity_learner.learn_alias(a_raw, b_raw)
            return True

        # 6. Fuzzy Similarity on Normalized Roots
        sim = string_similarity(a_clean, b_clean)
        if sim >= 0.82:
            vendor_identity_learner.learn_alias(a_raw, b_raw)
            return True

    return False
